### PR TITLE
Add response body in middlewares before and after request

### LINKED ISSUE #244: [Feature Request] Add response body in middlewares before and after request
**Current Behavior**

Now, there are four issues related to this topic:
- It should be possible to access `response["body"]` in the `after_request` middleware.
- There is no response dictionary accessible in the middleware. We only return the request object.
- If we have a `before_request` middleware before a middleware then we can't access the body in the post route.
- And finally, we can't even access the body in `before_request`

Try adding this in the `base_routes`:
```python3
@app.before_request("/post_with_body")
async def hello_before_request(request):
    global callCount
    callCount += 1
    print(request)
    return ""


@app.after_request("/post_with_body")
async def hello_after_request(request):
    global callCount
    callCount += 1
    print(request)
    return ""
```

**Desired Behavior**

Issues are fixed 😅 

---
**Screenshots / Mockups**

**Alternatives**

---

### PR DESCRIPTION
**Pull request by Shending-Help**
**Description**
This PR fixes the issue with request body not being accessible in the middlewares and add a new feature of having the response object accessible in the before and after request middlwares

This PR fixes #244

### CODE REVIEW

File: integration_tests/base_routes.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ async def hello(request):
     global callCount
     callCount += 1
     message = "Called " + str(callCount) + " times"
-    print(message, request)
+    # print(message, request)
```
Comment (sansyrox) [inline integration_tests/base_routes.py:48]: Can you please uncomment it?

File: integration_tests/base_routes.py [hunk 2]
```diff
@@ -74,20 +74,34 @@ def return_int_status_code():
...
     return ""
+    
+# @app.before_request("/")
+# async def hello_before_request(request, response):
+#     global callCount
+#     callCount += 1
+#     print()
+#     response["body"] = "hello before request"
+#     return ""
+
```
Comment (sansyrox) [inline integration_tests/base_routes.py:103]: Can you please remove it?

File: robyn/router.py [hunk 3]
```diff
@@ -81,8 +81,9 @@ def __init__(self) -> None:
         super().__init__()
         self.routes = []
 
-    def add_route(self, route_type: str, endpoint: str, handler: Callable) -> Callable:
-        number_of_params = len(signature(handler).parameters)
+
+    def add_route(self, route_type: str, endpoint: str, handler: Callable, number_of_params=0) -> None:
+        print(route_type, handler, signature(handler).parameters, number_of_params)
```
Comment (sansyrox) [inline robyn/router.py:86]: We shouldn't have print statements in internal functions.

File: robyn/router.py [hunk 4]
```diff
@@ -100,8 +101,9 @@ def add_route(self, route_type: str, endpoint: str, handler: Callable) -> Callab
     # and returns the arguments.
     # Arguments are returned as they could be modified by the middlewares.
     def add_after_request(self, endpoint: str) -> Callable[..., None]:
-        def inner(handler):
-            @wraps(handler)
+    
+        def inner(handler): # number_of_params
```
Comment (sansyrox) [inline robyn/router.py:105]: Extra comment.

File: src/executors/mod.rs [hunk 5]
```diff
@@ -7,50 +7,49 @@ use std::collections::HashMap;
 use std::rc::Rc;
 use std::sync::Arc;
 
+use actix_web::HttpResponse;
```
Comment (sansyrox) [inline src/executors/mod.rs:10]: There are a few unused imports here.

File: src/executors/mod.rs [hunk 6]
```diff
@@ -7,50 +7,49 @@ use std::collections::HashMap;
...
 use pyo3::prelude::*;
 use pyo3::types::PyDict;
 
 /// @TODO make configurable
-const MAX_SIZE: usize = 10_000;
 
 pub async fn execute_middleware_function<'a>(
     function: PyFunction,
-    payload: &mut web::Payload,
+    payload: &mut Vec<u8>,
```
Comment (sansyrox) [inline src/executors/mod.rs:26]: We can replace `&mut Vec<u8>` with `&mut [u8]` to make it more generic

File: src/executors/mod.rs [hunk 7]
```diff
@@ -7,50 +7,49 @@ use std::collections::HashMap;
...
     headers: &HashMap<String, String>,
     req: &HttpRequest,
```
Comment (sansyrox) [inline src/executors/mod.rs:28]: We can remove the `req` param.

File: src/executors/mod.rs [hunk 8]
```diff
@@ -7,50 +7,49 @@ use std::collections::HashMap;
...
-            body.extend_from_slice(&chunk);
-        }
-
-        data = body.to_vec()
+    let data = payload.clone();
+    let tmp = &HttpResponse::Ok().finish();
+
+    // make response object accessible while creating routes
+    let response = match res {
+        Some(res) => res.clone(),
```
Comment (sansyrox) [inline src/executors/mod.rs:42]: You can just return `res` here?

File: src/executors/mod.rs [hunk 9]
```diff
@@ -184,7 +198,7 @@ pub async fn execute_function(
 #[inline]
 pub async fn execute_http_function(
     function: PyFunction,
-    payload: &mut web::Payload,
+    payload: &mut Vec<u8>,
     headers: HashMap<String, String>,
     req: &HttpRequest,
```
Comment (sansyrox) [inline src/executors/mod.rs:203]: We can remove the `req` parameter from here.

**Review by sansyrox — COMMENTED** [review]
Thank you for the PR @Shending-Help 😄 

I have a few inline comments.

File: src/executors/mod.rs (hunk 7 above)
Comment (Shending-Help) [reply src/executors/mod.rs:28]: I've removed the req since its is not used but would it be better to make it ignored like this `_req` for any later use?

**Review by sansyrox — COMMENTED** [review]
@Shending-Help , can you please add one integration test to confirm the working of everything?

File: src/executors/mod.rs [hunk 10]
```diff
@@ -7,50 +7,47 @@ use std::collections::HashMap;
...
-            // limit max size of in-memory payload
-            if (body.len() + chunk.len()) > MAX_SIZE {
-                bail!("Body content Overflow");
-            }
-            body.extend_from_slice(&chunk);
-        }
-
-        data = body.to_vec()
+    let data = payload.to_owned();
+    let tmp = &HttpResponse::Ok().finish();
```
Comment (sansyrox) [inline src/executors/mod.rs:36]: can you please rename to `temp_response` or something a bit more meaningful than `tmp`?

File: src/executors/mod.rs [hunk 11]
```diff
@@ -69,14 +66,25 @@ pub async fn execute_middleware_function<'a>(
...
+                response_dict.insert("headers", response_headers.into_py(py));
+                response_dict.insert("status", response_status_code.into_py(py));
+                response_dict.insert("body", response_body.into_py(py));
+                info!("{:?}", request);
+                info!("{:?}", response_dict);
+                info!("{:?}", number_of_params);
+
+                // let response = handler.call1((request.clone(),));
+                // pyo3_asyncio::tokio::into_future(response?);
+                // response = response.await?;
```
Comment (sansyrox) [inline src/executors/mod.rs:79]: This is dead code here.

File: src/executors/mod.rs [hunk 12]
```diff
@@ -69,14 +66,25 @@ pub async fn execute_middleware_function<'a>(
...
                 // is this a bottleneck again?
                 request.insert("headers", headers.clone().into_py(py));
-                // request.insert("body", data.into_py(py));
+                request.insert("body", data.into_py(py));
+                response_dict.insert("headers", response_headers.into_py(py));
+                response_dict.insert("status", response_status_code.into_py(py));
+                response_dict.insert("body", response_body.into_py(py));
+                info!("{:?}", request);
+                info!("{:?}", response_dict);
+                info!("{:?}", number_of_params);
```
Comment (sansyrox) [inline src/executors/mod.rs:75]: Do we need these logs?

File: src/executors/mod.rs [hunk 13]
```diff
@@ -184,34 +196,33 @@ pub async fn execute_function(
...
+    //         let chunk = chunk?;
+    //         // limit max size of in-memory payload
+    //         if (body.len() + chunk.len()) > MAX_SIZE {
+    //             bail!("Body content Overflow");
+    //         }
+    //         body.extend_from_slice(&chunk);
+    //     }
+
+    //     data = body.to_vec()
+    // }
```
Comment (sansyrox) [inline src/executors/mod.rs:225]: dead code here

**Review by sansyrox — COMMENTED** [review]
I have some more inline comments on the PR. Thanks again 😄

File: src/executors/mod.rs (hunk 12 above)
Comment (Shending-Help) [reply src/executors/mod.rs:75]: no we don't these were just for debugging purposes

**Comment by sansyrox** [comment]
@Shending-Help , there are some conflicts now. Apologies .

File: src/server.rs [hunk 14]
```diff
@@ -392,35 +393,60 @@ async fn index(
...
+        }
+
+        data = body.to_vec()
+    }
 
+    // payload = [1,2,3,4,5,'\0']
+    //           i=0
+    //          .next() -> 1
+    //          i=1
+    // need a better name for this
```
Comment (sansyrox) [inline src/server.rs:421]: We have dead code here too.

File: src/server.rs [hunk 15]
```diff
@@ -392,35 +393,60 @@ async fn index(
...
+            )
+            .await;
+            debug!("Middleware contents {:?}", x);
+            x
+        }
+        None => HashMap::new(),
+    };
+
+
+    // payload = ['\0']
```
Comment (sansyrox) [inline src/server.rs:442]: extra comment.

File: src/server.rs [hunk 16]
```diff
@@ -392,35 +393,60 @@ async fn index(
...
 
+    // payload = [1,2,3,4,5,'\0']
+    //           i=0
+    //          .next() -> 1
+    //          i=1
+    // need a better name for this
+    let tuple_params = match middleware_router.get_route("BEFORE_REQUEST", req.uri().path()) {
+        Some(((handler_function, number_of_params), route_params)) => {
+            let x = handle_http_middleware_request(
+                // potentially return the data method
```
Comment (sansyrox) [inline src/server.rs:425]: and here too.

File: src/executors/mod.rs (hunk 10 above)
Comment (Shending-Help) [reply src/executors/mod.rs]: done

**Comment by Shending-Help** [comment]
removed all dead code and addressed all comments and resolved the merge conflicts

**Comment by sansyrox** [comment]
Thanks @Shending-Help . Great work 😄 

I will have a look tomorrow morning, Indian time.

File: src/executors/mod.rs (hunk 7 above)
Comment (sansyrox) [reply src/executors/mod.rs]: @Shending-Help , I have learned from my experience that it usually doesn't make sense to abstract for the future as much. If need be, we'll add the parameter again. 😄

**Comment by sansyrox** [comment]
@Shending-Help ,  can you please add one integration test too? I think you may have missed this comment. I'll still do a first review tomorrow morning 😄 

> @Shending-Help , can you please add one integration test to confirm the working of everything?

File: src/executors/mod.rs [hunk 17]
```diff
@@ -105,12 +106,16 @@ pub async fn execute_middleware_function<'a>(
...
+                response_dict.insert("status", response_status_code.into_py(py));
+                response_dict.insert("body", response_body.into_py(py));
 
                 let output: PyResult<&PyAny> = match number_of_params {
                     0 => handler.call0(),
                     1 => handler.call1((request,)),
+                    2 => handler.call1((request, response_dict)),
                     // this is done to accomodate any future params
-                    2_u8..=u8::MAX => handler.call1((request,)),
+                    3_u8..=u8::MAX => handler.call1((request, response_dict)),
```
Comment (sansyrox) [inline src/executors/mod.rs:118]: Everything looks great. Just one last nit:

`request` and `response_dict` do not match. Can you rename them to `request` and `response` or  `req` and `res` or `request_dict` and `response_dict`?

**Comment by sansyrox** [comment]
These have been implemented now. Hence, closing the PR

**Pull request closed without merging**